mlx/backend/common/qrf.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright © 2023 Apple Inc.
+
+#include <cassert>
+
+#include <iostream>
+#include "mlx/allocator.h"
+#include "mlx/backend/common/copy.h"
+#include "mlx/device.h"
+#include "mlx/dtype.h"
+#include "mlx/ops.h"
+#include "mlx/primitives.h"
+#include "mlx/utils.h"
+
+#ifdef ACCELERATE_NEW_LAPACK
+#include <vecLib/lapack.h>
+#include <vecLib/lapack_types.h>
+#endif
+
+namespace mlx::core {
+
+template <typename T>
+struct lpack;
+
+template <>
+struct lpack<float> {
+  static void xgeqrf(
```

Comment:
Also I'm curious, why is this "xgeqrf" instead of `sgeqrf` which is the reference in the docs for single precision:

https://www.netlib.org/lapack/single/sgeqrf.f

Reply:
internally it is calling sgeqrf. 

there are variants of 'sgeqrf', such as 'dgeqrf' for double precision and 'cgeqrf' for complex numbers. the above template, the 'x' in 'xgeqrf' serves as a placeholder that can be defined for a specific type T. 

basically, the same template can be used for different data types by simply substituting the appropriate type-specific LAPACK routine